### PR TITLE
Build with windows line endings

### LINKED ISSUE #6249: Segmentation fault when building Coq on Windows 10
#### Version

The Coq Proof Assistant, version 8.7.0 (November 2017)
compiled on Nov 26 2017 3:03:40 with OCaml 4.06.0
[I checked out v8.7.0 in git]

#### Operating system

Windows 10
i7-5500U processor (2 core, hyperthreaded)

#### Description of the problem

When I run 'make' to build Coq, I get a Segmentation Fault every time in theories/FSets/FMapFullAVL.v:

```
$ make VERBOSE=1
make --warn-undefined-variable --no-builtin-rules -f Makefile.build
make[1]: Entering directory '/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq'
rm -f theories/FSets/FMapFullAVL.glob
bin/coqtop.exe -boot   -compile theories/FSets/FMapFullAVL.v
make[1]: *** [Makefile.build:745: theories/FSets/FMapFullAVL.vo] Segmentation fault
make[1]: *** [theories/FSets/FMapFullAVL.vo] Deleting file 'theories/FSets/FMapFullAVL.glob'
make[1]: Leaving directory '/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq'
make: *** [Makefile:159: submake] Error 2
```

It blows up on line 70 (or just after line 70) in that file:
```
bin/coqtop.exe -boot   -compile-verbose theories/FSets/FMapFullAVL.v -time
     :
inv avl; intuition; omega_max.
Chars 2124 - 2154 [(inv~avl;~intuition;~omega_max).] Segmentation fault
```

Emilio suggested this may be the same bug as https://github.com/coq/coq/issues/5517.  However, my processor is an i7-5500U, in the Broadwell family while that bug only mentions the Kaby Lake and Skylake families.  Broadwell is the prior family.  I've not found anything about a problem in Broadwell or a microcode update from Intel for it.

"-native-compiler no" on the ./configure didn't help

I've not found a way to turn off hyperthreading for Windows 10.

I had to convert ./configure from DOS to Unix line endings.  Also kernel/copcodes.ml had a number of return characters that made that file not compile.  I removed those.  No other changes.

This is my first time trying to compile Coq and my first use of OCaml.

Here is my ./configure:

```
$ ./configure -native-compiler no                                                                                     
You have OCaml 4.06.0. Good!
You have Camlp5 7.03. Good!
You have native-code compilation. Good!
No such directory 'C:/OCaml64/home/Jim/.opam/4.06.0+mingw64c/lib/ocaml/lablgtk2' (in OCaml library).
LablGtk2 not found:
=> no CoqIde will be built.
Where should I install the Coq binaries [C:/coq/bin]?
Where should I install the Coq library [C:/coq/lib]?
Where should I install the Coqide configuration files [C:/coq/config]?
Where should I install the Coqide data files [C:/coq/share]?
Where should I install the Coq man pages [C:/coq/man]?
Where should I install the Coq documentation [C:/coq/doc]?
Where should I install the Coq Emacs mode [C:/coq/emacs]?
Where should I install the Coqdoc LaTeX files [C:/coq/latex]?

  Architecture                : win32
  Operating system            : Windows_NT
  Coq VM bytecode link flags  : -custom
  Other bytecode link flags   : -custom
  OS dependent libraries      : -cclib -lunix
  OCaml version               : 4.06.0
  OCaml binaries in           : C:/OCaml64/home/Jim/.opam/4.06.0+mingw64c/bin/
  OCaml library in            : C:/OCaml64/home/Jim/.opam/4.06.0+mingw64c/lib/ocaml
  OCaml flambda flags         :
  Camlp5 version              : 7.03
  Camlp5 binaries in          : C:\OCaml64\home\Jim\.opam\4.06.0+mingw64c\bin
  Camlp5 library in           : C:/OCaml64/home/Jim/.opam/4.06.0+mingw64c/lib/camlp5
  Native dynamic link support : true
  CoqIde                      : no
  Documentation               : None
  Web browser                 : start %s
  Coq web site                : http://coq.inria.fr/

  Native compiler for conversion and normalization disabled

  Paths for true installation:
  - the Coq binaries will be copied in C:/coq/bin
  - the Coq library will be copied in C:/coq/lib
  - the Coqide configuration files will be copied in C:/coq/config
  - the Coqide data files will be copied in C:/coq/share
  - the Coq man pages will be copied in C:/coq/man
  - the Coq documentation will be copied in C:/coq/doc
  - the Coq Emacs mode will be copied in C:/coq/emacs
  - the Coqdoc LaTeX files will be copied in C:/coq/latex

If anything is wrong above, please restart './configure'.

*Warning* To compile the system for a new architecture
          don't forget to do a 'make clean' before './configure'.
```

### PR DESCRIPTION
**Pull request by jfehrle**
These make it possible to run make successfully if the source tree was checked out with Windows line endings:
- Fixes #6249 Segmentation fault when building Coq on Windows 10
- Converts windows line endings before MD5 is computed to get the correct value

Plus:
- Remove incorrect quote marks in Windows build file (causes unknown parameter error in the next command)

(My first pull request for Coq.)

### CODE REVIEW

**Comment by Zimmi48** [comment]
Congrats on your first PR on Coq!

The proposed change seems rather simple and straightforward, however it creates an unnecessary dependency of the Makefile on `dos2unix`. While this command can easily be installed on Unix systems, it isn't necessary there so it seems a bit too much to add this to the list of requirements to build Coq. Could this step be made optional somehow?

I'd also like to understand what "the source tree was checked out with Windows line endings" exactly means. In your first reports, I seem to recall that you said you had first converted source files to use Windows line endings. Is this what this means? And why is this necessary?

Finally, just a small remark to say that I would like to wait for proper testing before merging this. In particular, this could affect the packaging script so testing this is dependent on solving #5998 first.

**Comment by JasonGross** [comment]
Do you need `dos2unix`, or is something like `sed 's/\r//'` enough?

**Comment by SkySkimmer** [comment]
It should be possible to use `tr` instead of dos2unix.

>I'd also like to understand what "the source tree was checked out with Windows line endings" exactly means. 

Probably autocrlf.

**Comment by Zimmi48** [comment]
Indeed @SkySkimmer. I didn't know about `git config --global core.autocrlf true`. This option seems immensely valuable for cross-platform collaboration so supporting it like this PR proposes is certainly a good idea.

Probably a good idea to use @JasonGross's or @SkySkimmer's suggestion to use `tr` which is part of coreutils, or `sed` instead of `dos2unix` to avoid a new dependency.

**Comment by herbelin** [comment]
By the way, is it intended that `"(* THIS FILE IS GENERATED. DON'T EDIT. *)"` is repeated for each new line of the source file? Shouldn't `awk` better be fed with a space-separated list free of any CR nor LF? (Or indeed use a simple OCaml-written translator instead.)

**Comment by Zimmi48** [comment]
Obviously this makes Travis fail with "dos2unix: command not found".
More interestingly this makes also AppVeyor (Windows tests) fail with:

```
File "kernel/cemitcodes.ml", line 165, characters 24-30:
Error: Unbound value opACC0
```

**Comment by jfehrle** [comment]
Thanks for your comments.

Good point about dos2unix, we don't want to use that.  I had tried to use sed before but had trouble finding the right parameters (the -e 's/\r//' doesn't work if it's last in the sed command line).

I agree with @herbelin that the repeated "THIS FILE IS GENERATED" messages are not needed but I didn't try to fix that because a) it is harmless, b) @maximedenes plans to rewrite the sed/awk in Caml and c) my lack of experience with sed/awk.

For the MD5, I put in something that works but it's a bit obscure.  The first sed removes the \r characters, then the second sed removes the " *-" that md5sum puts in the output when the input is from stdin.  If someone knows a cleaner way to do this, please let me know.

```
md5chk:
	$(SHOW)'MD5SUM cic.mli'
	$(HIDE)if grep -q `sed -n -e 's/\r//' -e '/^/p' <checker/cic.mli | $(MD5SUM) | sed -n -e 's/ \*\-//' -e '/^/p'` checker/values.ml; \
	       then true; else echo "Error: outdated checker/values.ml"; false; fi
```

The AppVeyor failure may well be from the "Fix build file" commit.  I will come back to that shortly.  I may need to remove that commit.

In git I revised the original commits but that seems to have created 2 additional commits with the same name and description.  Is that not going to work well in general?  What is the recommended approach for changing the code in a PR?

@Zimmi48 > I'd also like to understand what "the source tree was checked out with Windows line endings" exactly means. In your first reports, I seem to recall that you said you had first converted source files to use Windows line endings. Is this what this means? And why is this necessary?

Git did the conversion automatically.  I later ran dos2unix on a handful of files where the \r characters were a problem.

**Comment by SkySkimmer** [comment]
>For the MD5, I put in something that works but it's a bit obscure. The first sed removes the \r characters, then the second sed removes the " *-" that md5sum puts in the output when the input is from stdin. If someone knows a cleaner way to do this, please let me know.

bash has this feature:
```bash
echo <(echo foo)
# ==> /dev/fd/xxx
cat <(echo foo)
# ==> foo
```

**Comment by jfehrle** [comment]
Looks like a bunch of the jobs failed with "No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received".

What should I do about that?

**Comment by JasonGross** [comment]
Of the three that I picked somewhat randomly, they all failed in the same place:
```
COQC      theories/Reals/RiemannInt_SF.v
COQC      theories/Reals/RiemannInt.v
COQC      theories/Reals/Ranalysis5.v
COQC      theories/Reals/Ratan.v
COQC      theories/Reals/Machin.v
COQC      theories/Reals/Rtrigo.v
COQC      theories/Reals/Ranalysis.v
COQC      theories/Reals/NewtonInt.v
COQC      theories/Reals/Integration.v
COQC      theories/Reals/Reals.v
COQC      plugins/nsatz/Nsatz.v
COQC      plugins/setoid_ring/Rings_R.v
```

I'd be curious to know timings for the various files; do we have any in the standard library that take more than 10 minutes on travis?  (Perhaps we should be passing `TIMED=1` to building Coq?)  It seems possible but implausible that some files were already taking a very long time, and this changed the build order just enough to expose this issue.  I don't know what else it could be, though, unless something here makes the build infinite loop...

**Comment by Zimmi48** [comment]
> What is the recommended approach for changing the code in a PR?

The recommended approach is rebasing (to squash e.g. fix-up commits with the corresponding earlier commit) then force-pushing.

**Comment by jfehrle** [comment]
The Travis build behavior seems non-deterministic.  Perhaps too many jobs/not enough resources on worker nodes, so jobs are starved for CPU time?

Each worker hostname is a different physical machine (e.g. @1.production-4-worker-org-a-1-gce and @1.production-4-worker-org-b-1-gce) or are they multiple virtual machines sharing hardware?

**Comment by Zimmi48** [comment]
No, these failures can't be ignored. For some reason, this change makes `coqtop` hang when compiling the library. These failures cannot be observed on any other PR. Anyone has any idea what could be the cause of this phenomenon?

**Comment by SkySkimmer** [comment]
>@JasonGross noted a number of jobs failed on "COQC plugins/setoid_ring/Rings_R.v". When I run this command on my system, it completes in about 2 seconds. Why would the builds be likely to fail there?

Quite possibly it's failing on another step which is not the last run because of make parallelism.
Try modifying NJOBS to make it 1 in .travis.yml to disable the parallelism, maybe that will make it clearer.

>FWIW, I notice that job 3921.1 seems to be building much of coq a second time, just after "Running tests...", line 3949. That seems odd.

That's normal, it's because the test suite needs byte-compiled Coq.

**Comment by jfehrle** [comment]
I suspect Travis is having problems.  I will ask them.  I started a build around 1 PM Pacific time.  6 hours later none of the Travis jobs were running.  I submitted another build around 7 PM.  As of 11 PM Travis has not started (i.e. it lists 30 jobs but none have started).  AppVeyor looks like it may be complete (after lots running lots of test and a COQCHK--see last lines of the log below) but this page says it's not yet complete.

```
make[1]: Leaving directory '/cygdrive/c/projects/coq'
Build success
```

**Comment by SkySkimmer** [comment]
That's because all the PR travis jobs compete for resources (5 jobs at once per account). Enable travis for your own repository.

**Comment by jfehrle** [comment]
@SkySkimmer I enabled Travis for my repository, but doing a push triggered builds both in jfehrle/coq and coq/coq.  I don't need the latter.  Is there some option I need to change to avoid the duplicate builds?

**Comment by ejgallego** [comment]
@jfehrle I'd say don't worry, at most you can push a single build to Coq's master, and it is also useful.

**Comment by herbelin** [comment]
> Is there some option I need to change to avoid the duplicate builds?

I'd like too to have such option.

**Comment by Zimmi48** [comment]
> I'd like too to have such option.

We could easily implement a feature that does not launch a Travis job if e.g. the PR title contains "(no build)".

**Comment by jfehrle** [comment]
> We could easily implement a feature that does not launch a Travis job if e.g. the PR title contains "(no build)".

Or maybe the reverse--only do a build if "(do build)" is in the PR title.  It would be easy to forget to put in "(no build)", but readily noticeable if you leave out "(do build)"

**Comment by SkySkimmer** [comment]
>Is there some option I need to change to avoid the duplicate builds?

Don't push to the PR branch.

**Comment by Zimmi48** [comment]
> Don't push to the PR branch.

That's indeed a pretty basic solution :D

**Comment by jfehrle** [comment]
It took a while to figure this out.  It appears sed and bash don't behave exactly the same in Travis as on my Cygwin.  If I recall correctly (and simplified some):

- md5sum from a pipeline produces a string in the form "8e7cc35c798f7290748226596167a501 *-"
- locally, sed removed the trailing " *-", giving only the checksum.  But on Travis I got the checksum with a trailing "  -".
- substituting "\<checksum\> -" for $v in "grep -q $v checker/values.ml" produces 2 tokens to grep: \<checksum\> and "-".  "-" tells grep to read from stdin, which caused the make to hang.  If the string is only the checksum, it works fine.  The safe thing is to put $v in double quotes.
- in an alternate experiment, ${v/ */ checker/cic.mli} in the command line would replace the trailing " *-" with " checker/cic.mli", but Travis reported substitution errors.
- oddly enough, "md5sum <file" gives slightly different results than "cat file | md5sum".  The sum is the same but the output ends with "  -" for one and " *-" for the other.

**Comment by jfehrle** [comment]
To add to the fun, on OSX, sed doesn't understand \r to mean the return character, so "sed -n -e 's/\r//'" removes all the "r"s from the file, unlike Linux and Cygwin.  tr -d "\r" seems to work as expected on all 3 variants.

**Comment by jfehrle** [comment]
@Zimmi48 I have this working in my repository except for one failing job (https://travis-ci.org/jfehrle/coq/jobs/314178625).  Looks like it downloads some code from a couple git repositories.  The error message shows the compile is failing when trying to compile https://github.com/ppedrot/ltac2/blob/master/src/tac2core.ml.  What is the usual way to handle this kind of problem?  Should you or I ask @ppedrot to look at it/fix it?

```
CAMLOPT -c -for-pack Ltac2_plugin src/tac2core.ml
File "src/tac2core.ml", line 992, characters 17-41:
Error: Unbound constructor Genprint.TopPrinterBasic
Hint: Did you mean PrinterBasic?
make[2]: *** [src/tac2core.cmx] Error 2
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/travis/build/jfehrle/coq/_build_ci/coq-dpdgraph'
make: *** [ci-ltac2] Error 2
```

**Comment by SkySkimmer** [comment]
This is because some stuff (#6158) was merged in master breaking ltac2, so ltac2 changed as well breaking compatibility with old master which your branch is based on. You can rebase to test on your repository, but travis automatically merges with the target branch when testing PRs so you can also just push to this branch.

**Comment by jfehrle** [comment]
@SkySkimmer Thanks for the explanation.  The best practice is to base my changes off of "test"?  However, I don't understand how to "rebase to test on your repository" because I don't see a test branch in my repository (nor any obviously useful tags):

```
$ git branch -a
* (HEAD detached at origin/trunk)
  build_with_windows_line_endings
  master
  temp
  temp2
  temp3
  remotes/origin/HEAD -> origin/master
  remotes/origin/build_with_windows_line_endings
  remotes/origin/master
  remotes/origin/temp
  remotes/origin/temp2
  remotes/origin/temp3
  remotes/origin/trunk
  remotes/origin/v8.0
  remotes/origin/v8.1
  remotes/origin/v8.2
  remotes/origin/v8.3
  remotes/origin/v8.4
  remotes/origin/v8.5
  remotes/origin/v8.6
  remotes/origin/v8.7
```

**Comment by SkySkimmer** [comment]
"to" meaning "for that purpose"

Gaëtan Gilbert

**Comment by jfehrle** [comment]
Yes, not literally "test", but what do I rebase on to get the latest changes (if not master)?

**Comment by SkySkimmer** [comment]
I may have miscommunicated, you're supposed to rebase on master

Gaëtan Gilbert

**Comment by jfehrle** [comment]
Hmm.  Making some progress, but still some confusion.  I did a "git remote set-url origin https://github.com/jfehrle/coq.git" a couple days ago, so when I was doing "git pull master" before it would not pick up changes in coq.coq.

I just did a "git pull https://github.com/coq/coq.git master", which pulled a bunch of changes since Dec 1.  Oddly, some of the changes were applied after my changes (which I fixed).  I moved the temp3 and build_with_windows_line_endings branches to point to my changes and pushed each branch.

The builds are still in progress, but I see 31 jobs in my build https://travis-ci.org/jfehrle/coq/builds/314461826 (temp3 branch) but only 30 jobs in https://travis-ci.org/coq/coq/builds/314403530, so it seems they are processing somewhat different code.  I don't understand why this would be different.

**Comment by SkySkimmer** [comment]
One of the OSX jobs is used for packaging reasons and therefore is not 
run for PRs

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
You may have already figured this out but my personal workflow when I want to rebase is:

```
git checkout master
git pull
git checkout my-branch
git rebase master
```

For this to work, I need to have the `master` branch tracking the main GitHub repository. This is how I achieve this:

```
git remote add upstream https://github.com/coq/coq
git fetch upstream
git branch --set-upstream-to=upstream/master master
```

The `sed` issue reminds me that the Coq packaging scripts are using a different version of `sed` that the one that is shipped with Cygwin. @MSoegtropIMC is the author of those packaging scripts and might be able to say more. But this is not necessarily relevant to your issue.

**Comment by jfehrle** [comment]
Thanks, Théo.  At the top of this conversation page, I see "Commits 48" and "Files changed 3".  The commits are from rebasing--I hope they are harmless, and if not, let me know.  The 3 files changed are exactly what I want.

**Comment by SkySkimmer** [comment]
Please run `git rebase --onto 0275b58 56a70a9 build_with_windows_line_endings` (the first hash is current master, the second is the last unwanted commit).

**Comment by Zimmi48** [comment]
It is my understanding that this PR is ready as far as the author (@jfehrle) is concerned. Given that it affects the kernel, it should still get one or two formal approvals (@SkySkimmer @maximedenes @herbelin? anyone else?) and given that it affects the build, we should also test the Windows packaging scripts with this PR. @maximedenes any chance you can do that with your VM? Otherwise, I might try to find the time to do it...

File: Makefile.checker
Comment (SkySkimmer) [inline Makefile.checker]: Whichever character appears first after `s` is used as separator, a good pick avoids the need to escape `/`.
When `/` needs to appear in the expression I usually see `:` as separator, so 
`'s: .*: checker/cic.mli:'`

File: Makefile.checker
Comment (SkySkimmer) [inline Makefile.checker]: Maybe echo to stderr (`>&2`).

**Review by SkySkimmer — APPROVED** [review]
Comments are nits.

**Comment by jfehrle** [comment]
Yes, this is ready.  I added the changes SkySkimmer suggested.

**Comment by jfehrle** [comment]
The failure seems like it may be a Travis operational error.  I can't re-run failed jobs or builds for coq/coq from the Travis GUI (although I can for jfehrle/coq).  Do I have to do another push just to re-run it?

**Comment by Zimmi48** [comment]
No don't push

**Comment by Zimmi48** [comment]
I relaunched the failed job (these failures happen a lot, I don't know what is going on with Travis; FTR here is the corresponding issue https://github.com/travis-ci/travis-ci/issues/8507).
Given the Travis backlog these days, we want to avoid rebuilding everything for no good reason.

**Comment by Zimmi48** [comment]
Cherry-picking these two commits on v8.7 leads to a failure, see here https://travis-ci.org/Zimmi48/coq/jobs/316855931#L2434
Given that 8.7.1 is supposed to be released today, this is a bit late for debugging this so I propose postponing to 8.7.2 (i.e. backporting to v8.7 shortly after the 8.7.1 release) or 8.8 (i.e. merging this PR into master only).
Given that your use case is to be able to hack Coq after cloning it with git, I suppose it doesn't matter too much if the fix ends up in the 8.7.1 release or is merged shortly after.
BTW, to reproduce the issue just run `git checkout v8.7 && git cherry-pick ae460dc`, then build *on Linux* (note that the Windows subsystem for Linux should allow you to test what happens on Linux without too much hassle).

**Comment by jfehrle** [comment]
I will look at it.  But probably easier for me to build it/test it on Travis.

**Comment by jfehrle** [comment]
8.7.2 would be fine.  The checksum itself is correct, but in this branch there's an extra space between the checksum and "checker/cic.mli" in values.ml.  The fix (not tested) is to only look for the checksum itself and not for the filename.replacing a line in Makefile.checker:

```
	$(HIDE)v=`tr -d "\r" < checker/cic.mli | $(MD5SUM) `; \
```

**Comment by jfehrle** [comment]
I checked out v8.7.0 and cherry-picked updated code.  It works, except for one job https://travis-ci.org/jfehrle/coq/jobs/317222741 that failed with

```
Checking out /home/travis/build/jfehrle/coq/_build_ci/HoTT
+git fetch https://github.com/ejgallego/HoTT.git ocaml.4.02.3
fatal: Couldn't find remote ref ocaml.4.02.3
make: *** [ci-hott] Error 128
```

and another that timed out.

I also did a build on master with the changed code https://travis-ci.org/jfehrle/coq/builds/317166644 with 2 failures.  I don't think these failures are mine doing--what should I do about them?  Might as well fix these in my branch before rebuilding in the PR.

BTW, build_with_windows_line_endings has 2 commits; both should be cherry picked.

**Comment by Zimmi48** [comment]
OK, thank you @jfehrle for looking into this and for the fix.
Don't worry about the failures, I know exactly where they come from. I've launched a final test here https://travis-ci.org/Zimmi48/coq/builds/317323686 and here https://ci.appveyor.com/project/Zimmi48/coq/build/backport-pr-6305~215.
If they pass as expected (especially the second one, which will finally allow testing the package build scripts), I'll recommend merging this pull request immediately.
(Only the first commit needed to be backported, because the second one was also included in #6437.)

**Comment by Zimmi48** [comment]
OK all tests have passed, so this is ready to merge AFAICT.

**Comment by jfehrle** [comment]
If all the tests pass, then you should have the right changes, though I can't see them to doublecheck.

I just pushed the my final version to build_with_windows_line_endings.  Maybe that was not needed.

Thanks for your help.

**Comment by Zimmi48** [comment]
Indeed, that was not needed. What did you change in this final version?

**Comment by Zimmi48** [comment]
FTR here was my backport commit https://github.com/Zimmi48/coq/commit/77c1a5dc3fd25894bf2fc5e136e7761499706575, although I will have to redo it now.

**Comment by jfehrle** [comment]
The checksum was correct, but in this branch there is an extra space between the checksum and "checker/cic.mli" in values.ml. The fix is to grep for only the checksum itself but not also for the filename.  So now the sed in Makefile.checker is:

```sed -n -e 's/ .*//' -e '/^/p'```

and the full change to the file is

```
	$(HIDE)v=`tr -d "\r" < checker/cic.mli | $(MD5SUM) | sed -n -e 's/ .*//' -e '/^/p'`; \
	       if grep -q "$$v" checker/values.ml; \
	       then true; else echo "Error: outdated checker/values.ml: $$v" >&2; false; fi
```

**Comment by Zimmi48** [comment]
Oh I see. My backport was removing that extra space but the new version should be even better and easier to backport.

**Pull request merged**
